### DISCUSSION TITLE
Promote `zopen_test` into a new high-level verb

### DISCUSSION THREAD

**Discussion by v1gnesh** [zopencommunity/meta#624] [Ideas]
... called `zopen-test` (although `/run tests` is available for manual invocation).
This will help us get granular reporting for CI and also doing just one if only one is necessary. 

Further, what about wrapping the various tests up in [bats](https://github.com/bats-core/bats-core)?

Thoughts? :thinking:

**Comment by DevonianTeuchter** [comment]
Did start playing with shunit2 as it supports sh, bash, zsh (and ksh should anyone use that!) so supports all the shells on zos.  Seemed a bit more usable than bats tbh but not really had the time to investigate properly!

**Reply by v1gnesh** [reply]
With zsh coming to 3.1, IMO, we could be open to adding a dependency on bash (alongside curl and jq).
If z/OS comes with an old version of bash (can't remember if it does...?), then that's a +1 for it.
If z/OS support for bash has been upstreamed, that's one more reason to safely rely on it.
Someone else said elsewhere... using substrings, arrays, etc. make the scripts look a lot cleaner, cutting down lines of ` | cut -d '..`.

But apart from that, since you're the shell wizard here, I'm sure your choice will be more informed :smile:

**Reply by v1gnesh** [reply]
Also came across https://shellspec.info/, which has a comparison of shunit2, bats, and shellspec in there.

**Reply by IgorTodorovskiIBM** [reply]
+1 for a test library. I looked into bats previously here: https://github.com/ZOSOpenTools/metaport/issues/34. shunit2/shellspec seem comparable.

What do you see `zopen test` doing? Would it run the check step in `zopen build`?  Or would it run the tests which are part of metaport? 

I think integrating the tests into meta is probably a good idea, that way we can leverage the tests as a way to also validate their setup as well.

**Reply by v1gnesh** [reply]
Main thing I'm thinking of is just splitting `build_and_test.yml` into `build.yml` and `test.yml`, allowing slightly more granular pass/fail badges in PRs (where we see the build pass tick), and setting the stage for similar expansion (ex: unit test, sec test, whatever).

**Reply by v1gnesh** [reply]
> Would it run the check step in zopen build?

To start off, `zopen test` would just do what `zopen_test()` function in buildenv does today.
And `zopen build` _wouldn't_/_shouldn't_ do it anymore.
**EDIT:** Ah yes, totally forgot about check. Yup, just pulling that out.

> Or would it run the tests which are part of metaport?

Hmm... so a theoretical `zopen test` should run the port's tests as well as test meta... ?

**Reply by IgorTodorovskiIBM** [reply]
Typically the tests are integrated into the build. So you need to go through the step of bootstrap, configure, build, and then test. `zopen test` could do all of those and disable install at the end.

**Reply by v1gnesh** [reply]
Hmm.. I've only seen a handful of build instructions in the wild. 
They're all like 
Do `make x` to build
Do `make test` to run tests
Do `make release` to create releases
Do `make release-specific1` for create a specific item
etc.

So in that way, `make test` would expect the binary to be available somewhere (or perhaps it builds first, by having a pre-req on `build` as you say), and is just to test it out.
Keeping it separate also means we have the choice of testing only major releases, or some such.

**Reply by IgorTodorovskiIBM** [reply]
Most of the autotools based projects use `make check`. For what you're proposing, `make installcheck` may be better. I opened an issue on this awhile back: https://github.com/ZOSOpenTools/meta/issues/598.
However, I'm not sure at this point how many tools specify an installcheck target - https://www.gnu.org/software/automake/manual/html_node/Basic-Installation.html
